langroid/agent/base.py

```diff
@@ -1260,51 +1282,112 @@ def maybe_parse(tool: type[ToolMessage]) -> Optional[ToolMessage]:
                 #     msg in chat_doc.tool_messages):
                 #    chat_doc.tool_messages.remove(msg)
                 # if we can handle it, do so
                 result = self.handle_tool_message(msg, chat_doc=chat_doc)
+                if result is not None and isinstance(result, ChatDocument):
+                    return result
             else:
                 # else wrap it in an agent response and return it so
                 # orchestrator can find a respondent
-                result = self.create_agent_response(tool_messages=[msg])
-        elif isinstance(msg, (ChatDocument, str)):
-            result = msg
-        elif isinstance(msg, BaseModel):
-            result = msg.json()
+                return self.response_template(author_entity, tool_messages=[msg])
         else:
-            # last resort: use json.dumps() or str() to make it a str
-            try:
-                result = json.dumps(msg)
-            except Exception:
-                try:
-                    result = str(msg)
-                except Exception as e:
-                    logger.error(
-                        f"""
-                            Error converting msg handler result to str: {e}", 
-                            """,
-                        exc_info=True,
-                    )
-                    result = None
-        return result
+            result = to_string(msg)
+
+        return (
+            None
+            if result is None
+            else self.response_template(author_entity, content=result, content_any=msg)
+        )
+
+    def from_ChatDocument(self, msg: ChatDocument, output_type: Type[T]) -> Optional[T]:
+        """
+        Extract a desired output_type from a ChatDocument object.
+        We use this fallback order:
+        - if `msg.content_any` exists and matches the output_type, return it
+        - if `msg.content` exists and output_type is str return it
+        - if output_type is a ToolMessage, return the first tool in `msg.tool_messages`
+        - if output_type is a list of ToolMessage,
+            return all tools in `msg.tool_messages`
+        - search for a tool in `msg.tool_messages` that has a field of output_type,
+             and if found, return that field value
+        - return None if all the above fail
+        """
+        content = msg.content
+        if output_type is str and content != "":
+            return cast(T, content)
+        content_any = msg.content_any
+        if content_any is not None and isinstance(content_any, output_type):
+            return cast(T, content_any)
+
+        tools = self.get_tool_messages(msg, all_tools=True)
+
+        if get_origin(output_type) is list:
+            list_element_type = get_args(output_type)[0]
+            if issubclass(list_element_type, ToolMessage):
+                # output_type = List[<ToolMessage],
+                # i.e. list of ToolMessage-derived objects
+                return cast(T, tools)
```

Comment:
As in the next case, we might want to filter the tools by those which are instances of `output_type`

Reply:
Pushed this change, but the original may be more readable, let me know what you think

Reply:
Made this change